src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -334,6 +347,13 @@ bool IsValidTriggerCharacterForDocument(Document document, char triggerCharacter
                     }
                 }
 
+                if (commitCharacters.SetEquals(CompletionRules.Default.DefaultCommitCharacters))
+                {
+                    // Cache that we only had the default characters to avoid checking again.
+                    currentRuleCache.Add(item.Rules.CommitCharacterRules, ImmutableArray<string>.Empty);
+                    return null;
+                }
```

Comment:
i underwstand now why i don't like this method.  It effectively repeats the empty/null check three times.  we should invert all of this:

1. check if the value is not in the list.  
2. if it is not compute it and store in the list
3. finally, detrermien what to return.  if it's an empty list, return null, otherwise return the list